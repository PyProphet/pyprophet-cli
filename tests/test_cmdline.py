# encoding: utf-8
from __future__ import print_function
import pdb
import subprocess
import pytest
import os
import shutil
import collections
import tempfile

import pandas

pandas.options.display.precision=6
pandas.options.display.width=200


@pytest.fixture(scope="session")
def test_data_folder():
    here = os.path.dirname(os.path.abspath(__file__))
    return os.path.join(here, "data")


@pytest.fixture(scope="session", params=[1, 3])
def setup(test_data_folder, request):
    num_files = request.param
    tmpdir = tempfile.mkdtemp()
    data_folder = os.path.join(tmpdir, "data_folder")
    os.makedirs(data_folder)
    working_folder = os.path.join(tmpdir, "working_folder")
    for i in range(num_files):
        shutil.copy(os.path.join(test_data_folder, "test_data.txt"), os.path.join(data_folder, "data_%d.txt" % i))

    Setup = collections.namedtuple("Setup", "data_folder working_folder number_input_files")
    return Setup(data_folder, working_folder, num_files)


def test_check(setup):
    cmd = ("pyprophet-brutus check --data-folder %s" % setup.data_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0


def test_subsample(setup):
    cmd = ("pyprophet-brutus subsample --job-number 1 --job-count 1 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--ignore-invalid-scores "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --working-folder %s") % (setup.data_folder, setup.working_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.working_folder)
    assert len(files) == setup.number_input_files + 1

    files.sort()
    subsamples = pandas.read_csv(os.path.join(setup.working_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)

    cmd = ("pyprophet-brutus subsample --job-number 1 --job-count 2 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--ignore-invalid-scores "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --working-folder %s "
           "--local-folder %s") % (setup.data_folder, setup.working_folder, tempfile.mkdtemp())
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.working_folder)
    assert len(files) == 2 if setup.number_input_files == 1 else 3

    files.sort()
    subsamples = pandas.read_csv(os.path.join(setup.working_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)

    cmd = ("pyprophet-brutus subsample --job-number 1 --job-count 2 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--ignore-invalid-scores "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --working-folder %s") % (setup.data_folder, setup.working_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.working_folder)
    assert len(files) == 2 if setup.number_input_files == 1 else 3

    files.sort()

    subsamples = pandas.read_csv(os.path.join(setup.working_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)


def test_learn(setup, regtest):
    cmd = ("pyprophet-brutus learn --random-seed 43 --working-folder %s" % setup.working_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0

    files = sorted(os.listdir(setup.working_folder))

    for i, file_ in enumerate(files):
        path = os.path.join(setup.working_folder, file_)
        print(i, "%7d" % os.stat(path).st_size, file_, file=regtest)

    for name in ("weights.txt", "sum_stat_subsampled.txt"):
        df = pandas.read_csv(os.path.join(setup.working_folder, name), header=None, sep="\t")
        print(df, file=regtest)
        # print(open(os.path.join(setup.working_folder, name), "r").read(), file=regtest)


def test_score(setup, regtest):
    cmd = ("pyprophet-brutus apply_weights --job-number 1 --job-count 1 "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --working-folder %s") % (setup.data_folder, setup.working_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
