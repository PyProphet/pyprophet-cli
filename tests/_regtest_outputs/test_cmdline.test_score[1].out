 0  448165 result_folder_1/data_0_scored.txt
 1     759 result_folder_1/summary_stats.txt
 0  119715 work_folder/data_0_score_data.npz
 1    5638 work_folder/full_stat_subsampled.txt
 2      17 work_folder/invalid_columns.txt
 3     392 work_folder/score_columns.txt
 4 1049943 work_folder/subsampled_data_0.txt
 5     615 work_folder/sum_stat_subsampled.txt
 6     433 work_folder/weights.txt

   qvalue   svalue   TP  FP        TN         FN      FDR     sens   cutoff
0    0.00  0.02915    9   0  78.33333  299.66667  0.00039  0.02916  6.12497
1    0.01  0.87564  270   3  75.33333   38.66667  0.00996  0.87473  4.19544
2    0.02  0.93303  288   6  72.33333   20.66667  0.02042  0.93305  3.83993
3    0.05  1.00000  309  17  61.33333   -0.33333  0.05067  1.00000  3.26682
4    0.10  1.00000  307  36  42.33333    1.66667  0.10561  0.99460  2.61622
5    0.20  1.00000  291  72   6.33333   17.66667  0.19833  0.94276  1.24635
6    0.30  1.00000  309  78   0.33333   -0.33333  0.20241  1.00000 -2.10786
7    0.40      NaN  NaN NaN       NaN        NaN      NaN      NaN      NaN
8    0.50      NaN  NaN NaN       NaN        NaN      NaN      NaN      NaN
